### DISCUSSION TITLE
Options to keep drafts in production builds

### DISCUSSION THREAD

**Discussion by Airkro** [facebook/docusaurus#7995] [Ideas]
Feature request:

Add an option to the config file, to keep drafts in production builds.

Use case:

Add `draft: true` to some docs' front-matter, bundle two sites, one of them hides the drafts as production, and another shows them for staging/internal.

**Comment by slorber** [comment]
We plan to have some kind of `createFrontMatter()` API (https://github.com/facebook/docusaurus/issues/5568) so that you can customize/update the frontmatter that Docusaurus reads directly from the config file. And you could use env variables in this hook.

Would your use-case be solved by such an imaginary config?

```js
const showDrafts = process.env.SHOW_DRAFTS === "true";

const siteConfig = {
  markdown: {
    createFrontMatter: async ({
      plugin,
      filePath,
      defaultCreateFrontMatter // reuse existing utils parseFrontMatter
    }) => {
      const frontMatter = await defaultCreateFrontMatter(filePath);
      if (frontMatter.draft === true && showDrafts) {
        return {...frontMatter, draft: false}
      }
      return frontMatter,
    }
  }
};
```

```bash
SHOW_DRAFTS=true yarn docusaurus build 
SHOW_DRAFTS=false yarn docusaurus build 
```

**Reply by Airkro** [reply]
I can accept that. wait for `createFrontMatter` landing.

I don't want to change `frontMatter.draft`, some [use cases](https://docusaurus.io/feature-requests/p/passing-front-matter-to-remarkplugins) may need it. but I can do some hacking instead.

**Comment by Airkro** [comment]
`createFrontMatter` landing maybe take years (in 3.0), before that, I want to inject a webpack loader by `configureWebpack` to do this.

But `docusaurus` always inject `@docusaurus\mdx-loader` after my loader, which means `@docusaurus\mdx-loader` always runs before my loader, maybe https://github.com/facebook/docusaurus/issues/6370 (also 3.0) is can solve this.

Seems like every way has been blocked. 😖

**Reply by Josh-Cena** [reply]
See https://docusaurus.io/community/release-process:

> In practice, you can expect a new major version every 2–4 months.

**Reply by slorber** [reply]
I have no idea what you want to do this your Webpack loader, but I suggest to wait for the `createFrontMatter` hook. It's a high priority for me to ship this asap, as it's not too complex to build and unlock great flexibility. Expect it to land in canary/3.0-alpha soon

**Comment by Airkro** [comment]
Update: 

~~Eventually, I found a way to inject a custom webpack loader, but `docusaurus` ignored drafts before the loader ran.~~

Now I override this line:
https://github.com/facebook/docusaurus/blob/d8c72fb32d9b2179b096961f1fe1ff83ae492056/packages/docusaurus-plugin-content-docs/src/index.ts#L79
To:
```js
const env = process.env.INTERNAL ? process.env.NODE_ENV : 'development' as DocEnv; 
```

When bundling the site on staging mode, set `process.env.INTERNAL === true`, draft pages will not ignore.

**Reply by Airkro** [reply]
In docusaurus 3.0, hacking this function is enough:

https://github.com/facebook/docusaurus/blob/bda8f14f8925de83c5a4f0c6a1430430dbc82ba1/packages/docusaurus-utils/src/contentVisibilityUtils.ts#L28-L39

**Comment by Airkro** [comment]
I also use a remark plugin to add a draft admonition when bundling.

https://github.com/nice-move/remark-docusaurus/blob/0902b3311cdc285106de3c7d9013e240335459a5/lib/draft-admonition.mjs

Admonitions will show on draft pages, whether in dev mode or not.

But it can be remove when this feature landing: https://github.com/facebook/docusaurus/issues/8285

**Reply by Airkro** [reply]
New version for Docusaurus 3.0:

https://github.com/show-docs/remark-docusaurus/blob/722fbfd7eefe48ebce280ae137a71f9973d00578/lib/draft-admonition.mjs

**Comment by Airkro** [comment]
Related: https://docusaurus.io/feature-requests/p/passing-front-matter-to-remarkplugins

**Reply by Airkro** [reply]
In docusaurus 3.0, remark plugin can read frontmatter now.
